### PR TITLE
Add the ability to run the go linters directly in the devcontainer

### PR DESCRIPTION
**Pull request by FlorentClarret**
### What does this PR do?

Add the ability to run the go linters directly in the devcontainer. 

- If the devcontainer is not configured (no devcontainer.json file), it will be generated
- If the devcontainer is not up, we will spin it up
- Otherwise we will just run the linters

The very first job on a fresh container takes around 4 minutes. Subsequent runs take around 50 seconds on my machine.

The idea is simple: 
- I added a new parameter called `run_on` and that can be `linux` for now
- This parameter is read by a decorator
- If specified and you are on linux already, we run the linters locally
- Otherwise we spin up the container and forward the command to the container  

This decorator is generic, we could add it to other commands, such as the one to run the unit tests. We can also extend it in the future to run the command on anything (a windows VM, in the cloud or whatnot).

The main reason I used a decorator is to be non-intrusive for the original task. Here I only need to add a parameter to the task.

I also started to add some tests. We do not have a lot of them currently but I'll try to add more in the future

### Motivation

- First step to make it easier for people to run the linters locally on various OS, here Linux. We will try to add Windows soon too

### Additional Notes

This solution is not perfect yet because:
- Users still need to manually pull the docker image 
- The docker image is hardcoded

I'll try to improve this in future PRs, I do not want to end up with a massive PR. 

https://datadoghq.atlassian.net/browse/ADXT-184

### CODE REVIEW

**Comment by FlorentClarret** [comment]
Current dependencies on/for this PR:

* main:
  * **PR #24758** 👈

This comment was autogenerated by Freephite.

File: tasks/devcontainer.py [hunk 1]
```diff
@@ -166,3 +169,39 @@ def is_up(ctx) -> bool:
 def is_installed(ctx) -> bool:
     res = ctx.run("which devcontainer", hide=True, warn=True)
     return res.ok
+
+
+def run_on_devcontainer(func):
```
Comment (KevinFairise2) [inline tasks/devcontainer.py:174]: **👏 praise:** ‏
Wow :o using decorator for that is such a nice idea!

File: tasks/devcontainer.py [hunk 2]
```diff
@@ -166,3 +169,39 @@ def is_up(ctx) -> bool:
...
+                    setup(ctx, image="486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-devenv:1-arm64")
+
+                if not is_up(ctx):
+                    print(color_message("Starting the devcontainer...", "orange"))
+                    start(ctx)
+
+                print(color_message("Running the command in the devcontainer...", "orange"))
+                cli = DockerCLI(DEVCONTAINER_NAME)
+
+                cmd = ["inv"] + sys.argv[1:]
```
Comment (KevinFairise2) [inline tasks/devcontainer.py:198]: Can we have a dead loop here? Looks like we will call
`inv linter.go --platform linux` in the devcontainer, what happens in this case?

**Review by KevinFairise2 — APPROVED** [review]
LGTM! Some small questions.

I tested the invoke tasks with devcontainer. We just need to make the pull of the image easier but this is working so well!
I really like how you integrated it with the existing tasks by only adding one more flag to the task

**Comment by FlorentClarret** [comment]
> I have one extra question, would it work out of the box on Windows? To run test on linux from Windows. If not we might add an early exit when Windows people try to use `--platform` saying that we do not support Windows yet but we're working on it?

I'll double check with someone on Windows, but it should be working for Windows users. All the tools we use here (npm, devcontainer cli and docker) should be available

File: tasks/devcontainer.py [hunk 3]
```diff
@@ -1,24 +1,27 @@
...
+import sys
 from collections import OrderedDict
+from functools import wraps
 from pathlib import Path
 
 from invoke import task
 from invoke.exceptions import Exit
 
 from tasks.build_tags import build_tags, filter_incompatible_tags, get_build_tags, get_default_build_tags
+from tasks.commands.docker import DockerCLI
```
Comment (chouetz) [inline tasks/devcontainer.py:17]:
```suggestion
from tasks.commands.docker import DockerCLI, AGENT_REPOSITORY_PATH
```
I think you don't need to redefine it below, correct?

File: tasks/devcontainer.py [hunk 4]
```diff
@@ -167,3 +170,39 @@ def is_up(ctx) -> bool:
...
+    All you need to do is to decorate your task with this decorator and add a `platform` argument to your task.
+    """
+
+    @wraps(func)
+    def _run_on_devcontainer(ctx, *args, **kwargs):
+        if kwargs.get('platform'):
+            platform = kwargs['platform'].lower()
+            if platform == "linux" and py_platform.system().lower() != platform:
+                # If we choose to run them on linux, and we are not on linux already
+                if not file().exists():
```
Comment (chouetz) [inline tasks/devcontainer.py:187]: This is not related to this PR: what about a little more precise name instead of `file` (eg `devcontainer_file`). It could be slightly confusing

File: tasks/devcontainer.py [hunk 5]
```diff
@@ -167,3 +170,39 @@ def is_up(ctx) -> bool:
...
+                    print(color_message("Generating the devcontainer file to run the linter in a container.", "orange"))
+                    # TODO remove the hardcoded image and auto-pull it
+                    setup(ctx, image="486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-devenv:1-arm64")
```
Comment (chouetz) [inline tasks/devcontainer.py:190]: you can maybe save its value as a "global" (line 24) meanwhile, wdyt?

**Review by chouetz — APPROVED** [review]

File: tasks/devcontainer.py (hunk 5 above)
Comment (FlorentClarret) [reply tasks/devcontainer.py:190]: Agreed!

File: tasks/devcontainer.py (hunk 4 above)
Comment (FlorentClarret) [reply tasks/devcontainer.py:187]: I see your point but this `file` is already in the `devcontainer` module so that should be fine, users outside this module will use `devcontainer.file` and I think that's fine. wdyt?

File: tasks/devcontainer.py (hunk 3 above)
Comment (FlorentClarret) [reply tasks/devcontainer.py:17]: Yup!

File: tasks/agent.py [hunk 6]
```diff
@@ -127,6 +129,7 @@ def build(
     bundle=None,
     bundle_ebpf=False,
     agent_bin=None,
+    platform=None,  # noqa: U100. Used by the run_on_devcontainer decorator
```
Comment (pgimalac) [inline tasks/agent.py:132]: If I understand correctly if the platform argument is provided then the task will actually be ran in the devcontainer, what do you think of renaming this to `devc_platform` (or anything mentioning devcontainer really) ?
I feel like it would make it clearer for users, and it would limit the risks of providing the argument thinking it's just a generic `platform` arg

**Review by pducolin — APPROVED** [review]
Some missing steps on my side:
* ensure docker version is 26
* run go mod vendor
Other than that worked smoothly!

File: tasks/agent.py (hunk 6 above)
Comment (FlorentClarret) [reply tasks/agent.py:132]: What about `--run-on` so it opens the door to more stuff in the future, not just devcontainer?

Comment (pgimalac) [reply tasks/agent.py]: I think it's better yep
Can you add an explicit help message for this arg ? (possibly listing allowed values)

**Review by pgimalac — APPROVED** [review]
